### ISSUE TITLE
Variant recipes in the crafting calculator

### ISSUE THREAD

**Issue by xvwyh** [gw2efficiency/issues#1653]
As the crafting calculator evolves and more and more features are added (like buying items from vendors presented as crafting steps), and more varient recipes and ways of items acquisitions are added to the game, the lack of an option of choosing which recipe to follow starts resembling a noobtrap for people who aren't aware of their options. Here's a couple of examples:

---

### Variant vendor exchanges

![image](https://user-images.githubusercontent.com/43143822/209568169-6ed32cd1-a478-43d3-b8a8-027d49574de6.png)

I was shocked at the calculator suggesting that I spend HoT currency on something as cheap and easy to obtain as Obsidian Shards. [There's a ton of different exchange variants for Obsidian Shards](https://wiki.guildwars2.com/wiki/Obsidian_Shard), and sure, different people will have excess amounts of different currencies, but for a player who doesn't know about the plethora of options they have - this will look both daunting and confounding - why does this core-tyria recipe require me to grind expansion currency?

This desperately needs some sort of a very visible option selector. Choosing one recipe, even if it's the one that uses the currency with the lowest account value, will still be taking agency out of the users' hands, as different players may place different value on different currencies. I, for one, give no regards for karma or unbound magic, but will do everything in my power to avoid frivolously spending HoT map currencies, as grinding for those is a royal pain in the ass in my opinion.

### Variant mystic forge recipes

![image](https://user-images.githubusercontent.com/43143822/209568949-62ba2d77-a8d4-4dfd-858b-737b8f9bc3c4.png)

[Amalgamated Gemstone](https://wiki.guildwars2.com/wiki/Amalgamated_Gemstone) is an example of an item that can be mystic forged using different recipes with different input materials. The crafting calculator select the recipe that uses the cheapest materials - this is good. But this may lead the players into mistakenly thinking that this is the only way to obtain the item, and that they need to buy this specific material, when they may already have stacks upon stacks of other possible materials (which, sure, aren't the cheapest option right now, but they *already have them*) lying aimlessly in their bank.

If the player has selected the "Use own materials" option - crafting calculator should do a pass over all the possible recipe variants first to see which ones can be crafted using player's owned mats, and only after that resort to suggesting to buy the remaining mats for the cheapest recipe from TP. (Although in Amalgamated Gemstone's case this is further complicated by the ability to mix-and-match different materials, generating recipes for every possible combination may be excessive).

### Variant crafting recipes

![image](https://user-images.githubusercontent.com/43143822/209569714-cdf77ec7-30f4-4d3e-8542-7863e0283786.png)

Same as the above, but concerns a crafting discipline instead of MF, so mixing and matching of materials isn't an issue. The example here - [Mystic Curio](https://wiki.guildwars2.com/wiki/Mystic_Curio) - concerns most of gen 2 legendaries: the crafting calculator again chooses the recipe with the cheapest materials (good), but in doing so completely disregards that the player may already own the slightly-more-expensive materials for other recipe variants.

### Gen 2 legendary weapons

Most gen 2 legendaries can be crafted with either [Gift of Maguuma Mastery](https://wiki.guildwars2.com/wiki/Gift_of_Maguuma_Mastery) or [Gift of Desert Mastery](https://wiki.guildwars2.com/wiki/Gift_of_Desert_Mastery). Currently, the calculator ~~defaults to the usage of the recipe that involves the PoF gift~~ picks a random recipe on a per-item basis with no rhyme or reason ([The Shining Blade](https://gw2efficiency.com/crafting/calculator/a~0!b~1!c~0!d~1-81957), for example, uses PoF gift, but [Pharus](https://gw2efficiency.com/crafting/calculator/a~0!b~0!c~0!d~1-89854) - uses HoT gift), with no way to select the other one. Even if the player already has a different gift partly crafted (maybe even fully crafted) - the calculator will insist on using the recipe that it wants to, failing to track the player's progress.

**Comment by queicherius** [comment]
This has been on my radar for a long time but is currently not a priority.

Having a choice of recipe sounds great in theory but in most cases where a different choice would "save a noob" would have 0 effect, because that user group would just go with the recipe suggested to them regardless - I know that from the tickets we get when something goes wrong and they just go with what is suggested. This is why we focus on getting the initial tree correct for 99% of use-cases.

> Variant vendor exchanges

Not an issue once #1593 is fully implemented (and not really "shocking" a beta feature has some weird edge cases), since then it will pick the most optimal "merchant" as well. In this case, optimal would not be based on account value, but a value assigned by us for the "acquisition difficulty" of the currency.

> Variant mystic forge recipes / crafting recipes

Both of your examples (gemstones & curios) are such low impact (if you compare crafting the suboptimal recipe with bought components + selling the components that could have been used VS crafting the optimal recipe) that it's priority is lower than most other planned features.

> Gen 2 legendary weapons

This is the only one that comes up with any regularity, and a work-around for this already exists via the multi-recipe feature:
https://gw2efficiency.com/crafting/calculator/a~0!b~1!c~0!d~1-81812;1-82003;1-71820;1-73239

**Comment by xvwyh** [comment]
> but a value assigned by us for the "acquisition difficulty" of the currency.

Like I said, this will be highly subjective. A WvW player might be swimming in excess skirmish tokens, a PvE player wouldn't even know what that is. I can't agree with the idea of someone deciding that for everyone.

> Both of your examples (gemstones & curios) are such low impact (if you compare crafting the suboptimal recipe with bought components + selling the components that could have been used VS crafting the optimal recipe) that it's priority is lower than most other planned features.

Fair enough, I guess. I only use the crafting calculator for legendaries, so for me it impacts 100% of my use cases, but I dunno how an average user uses the calculator.

**Comment by queicherius** [comment]
Note for making the gift selectable:

- Force e.g. `Shooshadoo` to always use Maguuma gift
- Add fake item (e.g. `Shooshadoo (Desert)`) with its own recipe with a Desert gift
- Allow multiple top-level recipes to exist (but only those ones)

**Comment by Hatar22** [comment]
> Note for making the gift selectable:
> 
> * Force e.g. `Shooshadoo` to always use Maguuma gift
> * Add fake item (e.g. `Shooshadoo (Desert)`) with its own recipe with a Desert gift
> * Allow multiple top-level recipes to exist

that's what came up in my mind as well ;)

**Comment by bluenote73** [comment]
It feels related to me so I'm erring on the side of not creating a new issue - I would love to see a way of toggling cost off for subcomponents (or granular 'use own'). This can be used for example to plan a legendary assuming a future starter kit, but it can also be used to make the displayed cost of say Reinforced Boreal Trunk be realistic and not inflated because of eitrite. That has the side effect of then making the mats list relevant.

**Comment by queicherius** [comment]
> It feels related to me so I'm erring on the side of not creating a new issue - I would love to see a way of toggling cost off for subcomponents (or granular 'use own'). This can be used for example to plan a legendary assuming a future starter kit, but it can also be used to make the displayed cost of say Reinforced Boreal Trunk be realistic and not inflated because of eitrite. That has the side effect of then making the mats list relevant.

That's somewhat included in https://github.com/gw2efficiency/issues/issues/58

**Comment by queicherius** [comment]
@Ecmelt This one is a bit bigger, let's chat before you start with it.
